packages/next/next-server/server/image-optimizer.ts

```diff
@@ -0,0 +1,245 @@
+    res.statusCode = 400
+    res.end('"url" parameter cannot be an array')
+    return { finished: true }
+  }
+
+  let absoluteUrl: URL
+  try {
+    absoluteUrl = new URL(url)
+  } catch (_error) {
+    // url was not absolute so assuming relative url
+    try {
+      absoluteUrl = new URL(url, `${proto}://${host}`)
+    } catch (__error) {
+      res.statusCode = 400
+      res.end('"url" parameter is invalid')
+      return { finished: true }
+    }
+  }
+
+  if (!['http:', 'https:'].includes(absoluteUrl.protocol)) {
+    res.statusCode = 400
+    res.end('"url" parameter is invalid')
+    return { finished: true }
+  }
+
+  if (!server.renderOpts.dev && !domains.includes(absoluteUrl.hostname)) {
+    res.statusCode = 400
+    res.end('"url" parameter is not allowed')
+    return { finished: true }
+  }
+
+  if (!w) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) is required')
+    return { finished: true }
+  } else if (Array.isArray(w)) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) cannot be an array')
+    return { finished: true }
+  }
+
+  if (!q) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) is required')
+    return { finished: true }
+  } else if (Array.isArray(q)) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) cannot be an array')
+    return { finished: true }
+  }
+
+  const width = parseInt(w, 10)
+
+  if (!width || isNaN(width)) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) must be a number greater than 0')
+    return { finished: true }
+  }
+
+  if (!sizes.includes(width)) {
+    res.statusCode = 400
+    res.end(`"w" parameter (width) of ${width} is not allowed`)
+    return { finished: true }
+  }
+
+  const quality = parseInt(q)
+
+  if (isNaN(quality) || quality < 1 || quality > 100) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) must be a number between 1 and 100')
+    return { finished: true }
+  }
+
+  const { href } = absoluteUrl
+  const hash = getHash([CACHE_VERSION, href, width, quality, mimeType])
+  const imagesDir = join(distDir, 'cache', 'images')
+  const hashDir = join(imagesDir, hash)
+  const now = Date.now()
+
+  if (await fileExists(hashDir, 'directory')) {
+    const files = await promises.readdir(hashDir)
+    for (let file of files) {
+      const [filename, extension] = file.split('.')
+      const expireAt = Number(filename)
+      const contentType = getContentType(extension)
+      if (now < expireAt) {
+        if (contentType) {
+          res.setHeader('Content-Type', contentType)
+        }
+        createReadStream(join(hashDir, file)).pipe(res)
+        return { finished: true }
+      } else {
+        await promises.unlink(join(hashDir, file))
+      }
+    }
+  }
+
+  const upstreamRes = await fetch(href)
+
+  if (!upstreamRes.ok) {
+    server.logError(
+      new Error(`Unexpected status ${upstreamRes.status} from upstream ${href}`)
+    )
+    return { finished: true }
```

Comment:
We need to end the request here or else it will just hang leaving the connection open

Reply:
Fixed, thanks! 👍